Add option to shift the start hour of the day

Meisenburger13 opened the pull request:
Hey, I frequently create my daily note after midnight, which means I wasn't able to use the shortcut for opening today's note. I wanted an option to delay the start of the day and thought others might find it useful too. 

This change lets users decide whether they want to shift the hours of the day by up to 6 hours in both directions. For example: setting it to 5 means that creating a note at any point before 5am will actually create a note for the day before. On the other hand setting it to -5 would shift the hours in the other direction, creating a note for the next day starting from 7pm.

Let me know if there are any issues.

leenzhu commented:
Hi @Meisenburger13,

Thanks for sharing your work.

But in my opinion we should not change the meaning of `today`. Changing the time range represented by `today` may lead to unnecessary misunderstandings. And also when you need the really `today` you have to modify the offset configuration to meet your demands.

Perhaps it would be better to add a new menu for your usage scenario.

Anyway, the menu `Open Another day's Note` may work properly for selecting other dates when needed.

Meisenburger13 commented:
I see your point, but honestly for me it's more intuitive that 'today' ends when I go to sleep. This is a feature in the Loop Habit Tracker which I used for a while and since then I'm really a fan and wish every app had this.

But it's fine if you're not interested to have it in the plugin, I'm just going to continue to use my branch then :)
If you are interested in general but with a different implementation (you mentioned a new menu?), I'd be willing to change my pull request as well. If not you can just close it.

leenzhu commented:
Although I personally do not use this feature, as you mentioned, others may have a need for it. If you are willing to make the modification, Since it will not affect the existing functionality, and it will add a new feature, which might be a better choice. Thanks!

Meisenburger13 commented:
Okay great! Could you explain which modifications you want more clearly though? I'm not sure what you meant with adding a new menu, should I change the description or something else?

leenzhu commented:
I mean don't change the origin command 'openTodayNote`, just create a new command named `openFuzzyToday`(I'm not clear how to name it :-)), something like:

```js
		await joplin.commands.register({
			name: "openFuzzyTodayNote",
			label: "Open Fuzzy Today's Note",
			execute: async () => {
				const d = await getDateWithOffset();
				const note = await createNoteByDate(d);
				await joplin.commands.execute("openNote", note.id);
				await joplin.commands.execute('editor.focus');
			}
		});

                ...

		await joplin.views.menus.create('journal-menu', 'Journal', [
                        ...
			{ label: "Open Fuzzy Today's Note", commandName: "openFuzzyTodayNote", accelerator: "CmdOrCtrl+Alt+X" },
                        ...

		]);
                
```

And don't forget to update the codes related to link insertion.

Meisenburger13 commented:
Ah ok, now I understand what you mean. I can implement it that way but I'm not sure it's really the best option. 

I feel like most people would only need one of the two options depending on their preference. I would for example only assign a shortcut to the fuzzy option if I were to implement it that way, while others don't have any use for the fuzzy option. Having a shortcut for both would be unnecessary for most, since you still have the `Open another day's note` if you really need Today's note, and it could be confusing. Also both options overlap during most hours of the day, which makes it redundant most of the time. 

I feel like my implementation keeps it hidden more from users who don't need the offset, but gives the option for users who do. Maybe I could adjust the description a bit, to make it clearer and less confusing to users?

If you prefer your implementation with the extra menu item, I can still implement it that way though, just let me know.

leenzhu commented:
You never know how your users will end up using your software.

If I change the original feature of `today`, It might cause unnecessary trouble.

Some users might be curious and want to try out this feature, so they configure the time offset. Later, they might realize they don’t actually need it but forget to remove the configuration. Much later, during their continued use, they may notice that the time is always incorrect, but by then, they might have forgotten that they had configured the time offset in the first place. Well, I've done something that stupid before.

My philosophy is to provide all features to users and let them decide which ones to use. In my design, `today` and `another day` are two fundamental features, and their functionality is completely unambiguous. All other features are added as extra options.

So I have to say that forgive my persistence, but I hope to keep the existing feature unaffected.

Meisenburger13 commented:
Thank you for clarifying your reasoning, I will get the pull request ready next week :)

Meisenburger13 commented:
Hey, these are the changes I've made:

I added an option with offset for every shortcut / menu item that uses `Today`. The default shortcuts are the same to the counterparts without an offset with an additional shift. 
I also changed the setting for the offset from advanced to default since a lot of the features now have an option with the offset, so the setting for it shouldn't be hidden from the user.

 I added information about the usage of the offset to the ReadMe file. 

Something we didn't discuss is the offset in the calendar. I thought it was more consistent and I find it less confusing to have the calendar highlight "today" with an offset as well, so I added a setting for that. I'm not sure why there's still a grey shadow on the actual today when doing so, but I don't think it's too distracting. 

Please let me know if you see any issues or have any feedback, especially on the calendar highlight since we didn't discuss that beforehand.

leenzhu commented:
Thanks for accepting my suggestions.

I have two points of disagreement with you.

First, since it is an `extra` feature for journal, the setting should be placed in the advanced section.

Second, please don't modify the calendar, the `today` in calendar is always presenting the real `today`. When user need the `offsetToday` they should not open the calendar. The only way to use `offsetTody` is through the newly added menu( or shortcut).

Here's another small suggestion: please add a separator to the menu.

Meisenburger13 commented:
Thanks for taking a look at the pull request and for your suggestions. 

I want to explain my reasoning for the offset in the calendar a bit more. I sometimes open the calendar after midnight and for me intuitively it still feels like the same day. So if I want to open the note from two days ago and i know today was a Tuesday (even though officially it's now a Wednesday), it happens that I will open Monday's note instead of Sunday's, since the calendar is highlighting Wednesday as `Today`.  It's an extra step for me to have to think about whether or not it's already after midnight and I have to offset my "two days ago" calculation.
I will keep that the offset for the calendar in my version, but if you don't want to add the feature I'll take it out.

leenzhu commented:
Thanks for your explanation about the offset in calendar.

The calendar is designed to choose the date as the absolute date no matter what date it is today, while you are using it for choosing relative date. Actually what you need is a relative date picker like `2 days ago`, `3 days later` rather than the normal calendar, at least not this kind of calendar.

I’d like to stand by my opinion on this matter.


The following link may help for creating a menu separator:

https://joplinapp.org/api/references/plugin_api/interfaces/menuitem.html

leenzhu commented:
Sorry, my fault, I just tried it and it didn't work. I didn't have it verified, and after reading the documentation, I took it for granted that it would work.

After reading the code of Joplin, I found that there is no relevant processing function. It’s possible that this feature has not been implemented by the official team.
